src/main/scala/is/hail/methods/LDMatrix.scala

```diff
@@ -23,10 +24,15 @@ object LDMatrix {
     
     val nSamples = vds.nSamples
 
-    val filteredNormalizedHardCalls = vds.rdd.flatMap { 
-      case (v, (_, gs)) => RegressionUtils.normalizedHardCalls(gs, nSamples).map(x => (v, x))
-    }
-      .persist()
+    val rowType = vds.rowType
+
+    val filteredNormalizedHardCalls = vds.rdd2.mapPartitions { it =>
+      val view = HardCallView(rowType)
+      it.flatMap { r =>
+        val v = UnsafeRow.readStruct(rowType, r.region, r.offset).getAs[Variant](1)
```

Comment:
I think this is not preferred going forward.  I'd probably write `Variant.fromRegionValue(r.region, rowType.loadField(r, 0))` (moving the necessary logic from `UnsafeRow.readStruct`.)

Reply:
I'm trying to write this, and it's horrible:
```scala
  def fromRegionValue(rv: RegionValue, offset: Long): Variant = {
    val t = TVariant.representation
    val altsType = t.fieldType(3).asInstanceOf[TArray]

    val contig = TString.loadString(rv.region, t.fieldOffset(offset + rv.offset, 0))
    val pos = rv.region.loadInt(t.fieldOffset(offset + rv.offset, 1))
    val ref = TString.loadString(rv.region, t.fieldOffset(offset + rv.offset, 2))

    val altsOffset = t.fieldOffset(offset + rv.offset, 3)
    val nAlts = altsType.loadLength(rv.region, altsOffset)
    val altArray = new Array[AltAllele](nAlts)
    var i = 0
    while (i < nAlts) {
      val o = altsType.elementOffset(altsOffset, nAlts, i)
      altArray(i) = AltAllele.fromMemoryBuffer(rv.region, o)
      i += 1
    }
    Variant(contig, pos, ref, altArray)
  }
```
This has a bug, but it's very difficult to reason about where. Am I missing something, or do we _really_ need better approaches to this kind of problem?

Reply:
Never use `loadOffset`.  That's probably your problem.  We should probably remove it or make it private to avoid confusion.  Use `loadField`.  `loadField` gives you the offset of the value of the field.  `loadOffset` gives you the offset of the field.  Those aren't the same for types that are stored as pointers in fields (arrays, strings, etc.)

Reply:
We should probably start writing some internal documentation about these things. I worked on these a few months ago and already found myself completely baffled.

Reply:
I agree completely.  I will start working on this.